File: Redundant-Connection.java
1class Solution {
2    public int[] findRedundantConnection(int[][] edges) 
3    {
4        DSU dsu = new DSU(edges.length);
5        int[] ans= new int[2];
6        for(int i=0;i<edges.length;i++)
7        {
8            if(!dsu.union(edges[i][0]-1, edges[i][1]-1))
9            {
10                ans[0] = edges[i][0];
11                ans[1] = edges[i][1];
12                break;
13            }
14            
15        }
16        return ans;
17    }
18    
19}
20
21class DSU
22{
23    int[] parent;
24    int[] rank;
25    DSU(int n)
26    {
27        parent = new int [n];
28        rank = new int[n];
29        for(int i=0;i<n;i++)
30            parent[i] = i;
31
32
33    }
34    int find(int x)
35    {
36        if(parent[x] != x)
37            parent[x] = find(parent[x]);
38
39        return parent[x];
40    }
41
42    boolean  union(int x, int y)
43    {   
44        int px = find(x), py = find(y);
45        if(px == py)
46            return false;
47        if(rank[px] > rank[py])
48            parent[py] = px;
49        else if(rank[py] > rank[px])
50            parent[px] = py;
51        else
52        {
53            parent[py] = px;
54            rank[px]++;
55        }
56        return true;
57    }
58
59
60
61}
62

File: Redundant-Connection.cpp
1class DSU
2{
3    public:
4    vector<int> parent;
5    vector<int> rank;
6    DSU(int n)
7    {
8        for(int i=0;i<=n;i++)
9        {
10            parent.push_back(i);
11            rank.push_back(1);
12        }
13    }
14    int find(int x)
15    {
16        if(parent[x]!= x)
17        {
18            parent[x] = find(parent[x]);
19        }
20        return parent[x];
21    }
22    bool _union(int x, int y)
23    {
24        int px = find(x), py= find(y);
25
26        if(px == py)
27            return false;
28        if(rank[px] > rank[py])
29        {
30            parent[py] = px;
31            
32        }
33        else if(rank[px] < rank[py])
34        {
35            parent[px] = py;
36            
37        }
38        else
39        {
40            parent[py] = px;
41            rank[px]++;
42           
43        }
44        return true;
45    }
46};
47class Solution {
48public:
49    vector<int> findRedundantConnection(vector<vector<int>>& edges) 
50    {
51        DSU dsu(edges.size());
52        vector<int> ans;
53        for(int i=0;i<edges.size();i++)
54        {
55            if(dsu._union(edges[i][0], edges[i][1]))
56            {
57                // do nothing
58            }
59            else
60            {
61                ans = {edges[i][0], edges[i][1]};
62            }
63        }
64        return ans;
65    }
66};
67